### ISSUE TITLE
The PRINCIPAL_NAME_INDEX_NAME key is not created in redis

### ISSUE THREAD

**Issue by wangtora** [redisson/redisson#2153]
What I want to do is depending on PRINCIPAL_NAME_INDEX_NAME in redis to control the Spring Security's maximum session management through SpringSessionBackedSessionRegistry class.

The first look at RedissonSessionRepository.save() confused me that it is an empty method which says "session changes are stored in real-time". In comparison, the standard RedisOperationsSessionRepository.save() of spring-session-data-redis has codes to create PRINCIPAL_NAME_INDEX_NAME. In fact, these codes are in RedisOperationsSessionRepository$RedisSession.saveDelta().

### Expected behavior
The PRINCIPAL_NAME_INDEX_NAME key is created with Spring Security's principal information

### Actual behavior
The PRINCIPAL_NAME_INDEX_NAME key is not created

### Steps to reproduce or test case
Using Spring Boot 2.1.3/Spring Security 5.1.4/Spring Session 2.1.4, during which the Spring Session java configuration is:
```java
@EnableRedissonHttpSession(keyPrefix = "xxx:")
public class SessionConfig {
    @Bean
    public static ConfigureRedisAction configureRedisAction() {
        return ConfigureRedisAction.NO_OP;
    }
}
```

### Redis version
5.0.0 on AWS Elasticache
### Redisson version
3.11.0
### Redisson configuration
```yaml
clusterServersConfig:
   retryAttempts: 6
   retryInterval: 20000
   readMode: "MASTER_SLAVE"
   nodeAddresses:
   - "redis://xxx.xxxxxx.clustercfg.apne1.cache.amazonaws.com:6379"
   scanInterval: 5000
```

**Comment by hs20xqy** [comment]
I had the same problem.I found RedissonSessionRepository.RedissonSession setAttribute method are interested in the redis writes PRINCIPAL_NAME_INDEX_NAME,but the code
` if (attributeName. equals (PRINCIPAL_NAME_INDEX_NAME) | | attributeName. equals (securityPrincipalSessionAttr)) `
always false, I don't confirm this is my configuration errors or bugs, my solution is to modify the source code,
change the code
`String securityPrincipalSessionAttr = getSessionAttrNameKey (SPRING_SECURITY_CONTEXT);`
to
 `String securityPrincipalSessionAttr = SPRING_SECURITY_CONTEXT;`
Now,PRINCIPAL_NAME_INDEX_NAME key is create in redis.
I'm not sure if this change will cause any other problems, I haven't found it yet.
Hope helpful!

**Comment by wangtora** [comment]
@mrniko Thanks for adding this issue to the 3.11.1 milestone, which may be fixed as @hs20xqy said. But 1 thing I must point out is that, after The PRINCIPAL_NAME_INDEX_NAME key is added, it cannot be deleted when the expired event happened. See the code in RedissonSessionRepository.onMessage():
```java
            String id = body.split(keyPrefix)[1];
            RedissonSession session = new RedissonSession(id);
            if (session.load()) {
                session.clearPrincipal();
            }
            publishEvent(new SessionExpiredEvent(this, session));
```

The if condition will always be false, as the expired session's key has already been deleted by redis at this moment. Nevertheless, the remaining principal won't cause a problem, if you use Spring Security's maximum session management in non-exception mode. It's just a waste of space. I know it's not easy to fix, but it's annoying in my firm even for this trival waste, so I am working on using Redisson without RedissonSessionRepository, as I said in #1897.

**Comment by mrniko** [comment]
Fixed both issues.

**Comment by wangtora** [comment]
@mrniko This commit 8d78bf5, caused other problems, during which the 1st is that, the expired session information cannot be deleted. Since the ttl has changed to another key(`<prefix>:sessions:expires:<sid>`), the session information(whose key is `<prefix>:<sid>`) without ttl is not deleted according to the following code in RedissonSessionRepository.onMessage():
```java
            String id = body.split(getExpiredKeyPrefix())[1];
            MapSession mapSession = loadSession(id);
            if (mapSession != null) {
                RedissonSession session = new RedissonSession(mapSession);
                session.clearPrincipal();
                publishEvent(new SessionExpiredEvent(this, session));
            }
```

There is no obvious code to delete key `<prefix>:<sid>`. I use redission 3.11.1-SNAPSHOT with redisson-spring-data-21 3.11.0, as redisson-spring-data-21 3.11.1-SNAPSHOT has compile error right now. Still working on other problems.

**Comment by mrniko** [comment]
@wangtora 

>  during which the 1st is that, the expired session information cannot be deleted

Thanks for pointing out! Fixed it here https://github.com/redisson/redisson/commit/e356c1e3d4b8d3787a6876a3aaf1157e953fb835

**Comment by wangtora** [comment]
Yes, it is fixed, and no more issues now, waiting for 3.11.1 release.
